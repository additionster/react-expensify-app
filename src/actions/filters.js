export const setTextFilter = (text = '') => ({
    type: 'SET_TEXT_FILTER',
    text
});

export const sortByAmount = () => ({
    type: 'SORT_BY_AMOUNT',
    sortBy: 'amount'
});

export const sortByDate = () => ({
    type: 'SORT_BY_DATE',
    sortBy: 'date'
});

export const setStartDate = (value = undefined) => ({
    type: 'SET_START_DATE',
    value
});

export const setEndDate = (value = undefined) => ({
    type: 'SET_END_DATE',
    value
});