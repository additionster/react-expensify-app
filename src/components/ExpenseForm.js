import React from 'react';
import moment from 'moment';
import DatePicker from 'react-datepicker';


class ExpenseForm extends React.Component {
    state = {
        description: this.props.expense? this.props.expense.description: '',
        amount: this.props.expense? (this.props.expense.amount/100).toString():'',
        note: this.props.expense? this.props.expense.note:'',
        createdAt: this.props.expense? moment(this.props.expense.createdAt):moment(),
        error: ''
    };
    onDescriptionChange = (e) => {
        const desc = e.target.value;
        this.setState(() => ({description:desc}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onCreatedAt = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }   
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.'}));
        }
        else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <div>
                {this.state.error && <p style={{color:"red"}}>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder='Description'
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text"
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <DatePicker 
                        showIcon
                        showMonthDropdown
                        showYearDropdown
                        dropdownMode='select'
                        selected={this.state.createdAt}
                        minDate={moment().subtract(10,'days').calendar()}
                        dateFormat="dd/MM/yyyy"
                        placeholderText='dd/MM/yyyy'
                        onChange={this.onCreatedAt}
                    />
                    <textarea 
                        value={this.state.note}
                        placeholder='Add a note for your expense  (optional)'
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>{this.state.description? "Edit":"Add"} Expense</button>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;