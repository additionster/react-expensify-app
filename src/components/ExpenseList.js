import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses && props.expenses.length !== 0? (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense}/>
                })
            ) : (
                <p>No expenses</p>
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};
//Connect store from redux using HOC pattern
const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseList;