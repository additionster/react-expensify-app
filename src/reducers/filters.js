import moment from 'moment';

const filterReducerDefaultState = {
    text: '',
    sortBy: 'date',
    startDate: moment().startOf('month'),
    endDate: moment().endOf('month')
};
const filterReducer = (state = filterReducerDefaultState, action) => {
    switch(action.type)
    {
        case 'SET_TEXT_FILTER':
            return {...state, text: action.text};
        case 'SORT_BY_AMOUNT':
        case 'SORT_BY_DATE':
            return {...state, sortBy: action.sortBy};
        case 'SET_START_DATE':
            const startDate = action.value? moment(action.value): action.value;
            return {...state, startDate };
        case 'SET_END_DATE':
            const endDate = action.value? moment(action.value): action.value;
            return {...state, endDate };
        default:
            return state;
    }
};

export default filterReducer;