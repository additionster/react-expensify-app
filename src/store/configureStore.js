import { configureStore, combineReducers } from "@reduxjs/toolkit";
import expenseReducer from "../reducers/expenses";
import filterReducer from "../reducers/filters";

export default () => {
    const store = configureStore({
        reducer: combineReducers({
            expenses: expenseReducer,
            filters: filterReducer
        }),
        middleware: (getDefaultMiddleware) => getDefaultMiddleware({
            serializableCheck: false
        }),
        devTools: window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
    });
    return store;
};
