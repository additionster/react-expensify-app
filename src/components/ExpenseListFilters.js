import React from "react";
import { connect } from "react-redux";
import { setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate } from "../actions/filters";
import DatePicker from "react-datepicker";

export class ExpenseListFilters extends React.Component {

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        switch(e.target.value)
            {
                case 'date': this.props.sortByDate();break;
                case 'amount': this.props.sortByAmount();break;
            }
    }

    onDateChange = (dates) => {
        const [startDate, endDate] = dates;
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    render() {
        return (
            <div>
                <input type="text" defaultValue={this.props.filters.text} onChange={this.onTextChange} />
                <select defaultValue={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DatePicker 
                    dateFormat='dd/MM/yyyy'
                    selectsRange={true}
                    startDate={this.props.filters.startDate?.toDate()}
                    endDate={this.props.filters.endDate?.toDate()}
                    onChange={this.onDateChange}
                    placeholderText="startDate - endDate"
                    isClearable={true}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (value) => dispatch(setTextFilter(value)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))

});
const ConnectedExpenseListFilters = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;