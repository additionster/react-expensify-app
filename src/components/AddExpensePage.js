import React from "react";
import { connect } from "react-redux";
import ExpenseForm from './ExpenseForm';
import { addExpense } from "../actions/expenses";
import { Navigate } from "react-router";


export class AddExpensePage extends React.Component {
    state = {
        submitted: false
    }
    onSubmit = (expense) => {
        //Cannot use hook in arrow function. so replace with <Navigate to/>
        //const navigate = useNavigate();
        this.props.addExpense(expense);
        //navigate('/');
        this.setState({submitted: true});
    }

    render() {
        return (
            <div>
                {this.state.submitted && <Navigate to="/" />}
                <h1>Add Expense</h1>
                <ExpenseForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    };
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);