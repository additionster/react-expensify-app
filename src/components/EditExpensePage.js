import React, { Component } from "react";
import { connect } from "react-redux";
import ExpenseForm from './ExpenseForm';
import { Navigate, useParams } from "react-router";
import { editExpense, removeExpense } from "../actions/expenses";

//wrap hook (useParams) inside function based component
function withParams(Component) {
    return props => <Component {...props} params={useParams()} />;
}

export class EditExpensePage extends React.Component {
    state = {
        submitted: false
    }

    onSubmit = (expense) => {
        this.props.editExpense(this.props.params.id, expense);
        this.setState({submitted: true});
    }

    onRemove = () => {
        this.props.removeExpense(this.props.params.id);
        this.setState({submitted: true});
    }

    render() {
        return (
            <div>
                {this.state.submitted && <Navigate to="/" />}
                <h1>Edit Expense</h1>
                <ExpenseForm 
                    expense={this.props.expenses.find((expense) => expense.id === this.props.params.id)}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>
                    Remove
                </button>
            </div>
        );
    };
}

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses
    };
};

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense({id}))
})

const connectedEditExpense = connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

export default withParams(connectedEditExpense);