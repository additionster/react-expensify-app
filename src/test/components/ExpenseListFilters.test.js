import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import '@testing-library/jest-dom';
import { BrowserRouter } from "react-router";
import configureStore from "../../store/configureStore";
import { Provider } from "react-redux";
import moment from "moment";
import { ExpenseListFilters } from "../../components/ExpenseListFilters";
import {filters, altFilters} from '../fixtures/filters';

let setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate, store, renderer;

beforeEach(()  => {
    setTextFilter = jest.fn();
    sortByDate = jest.fn();
    sortByAmount = jest.fn();
    setStartDate = jest.fn();
    setEndDate =jest.fn();
    store = configureStore();
    renderer = render(
        <ExpenseListFilters 
            filters={filters}
            setTextFilter={setTextFilter}
            sortByDate={sortByDate}
            sortByAmount={sortByAmount}
            setStartDate={setStartDate}
            setEndDate={setEndDate}
        />
    );
});

test('testRenderExpenseListFilters', () => {
    expect(renderer.baseElement).toMatchSnapshot();
});

test('testRenderExpenseListFiltersWithAltFilters', () => {
    const { rerender } = renderer;
    //re-render the component 
    rerender(<ExpenseListFilters filters={altFilters}/>);
    expect(renderer.baseElement).toMatchSnapshot();
});

test('testRenderExpenseListFiltersHandleTextChage', () => {
    const updatedText = 'Rent';
    const textInput = screen.getAllByRole('textbox').at(0);
    fireEvent.change(textInput, {
        target: { value: updatedText}
    });
    expect(setTextFilter).toHaveBeenCalledWith(updatedText);
});

test('testRenderExpenseListFiltersSortByDate', () => {
    const { rerender } = renderer;
    //re-render the component 
    rerender(<ExpenseListFilters filters={altFilters} sortByDate={sortByDate}/>);
    const selectInput = screen.getByRole('combobox');
    fireEvent.change(selectInput, {
        target: { value: 'date'}
    });
    expect(sortByDate).toHaveBeenCalled();
});

test('testRenderExpenseListFiltersSortByAmount', () => {
    const selectInput = screen.getByRole('combobox');
    fireEvent.change(selectInput, {
        target: { value: 'amount' }
    });
    expect(sortByAmount).toHaveBeenCalled();
});

test('testRenderExpenselistFiltersHandleDateChange', () => {
    const dateInput = screen.getAllByRole('textbox').at(1);
    fireEvent.change(dateInput, {
        target: { value: '01/01/2020 - 01/02/2020' }
    });
    expect(setStartDate).toHaveBeenLastCalledWith(moment('2020-01-01').toDate());
    expect(setEndDate).toHaveBeenLastCalledWith(moment('2020-02-01').toDate());
});