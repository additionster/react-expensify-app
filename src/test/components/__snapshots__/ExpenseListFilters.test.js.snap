// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`testRenderExpenseListFilters 1`] = `
<body>
  <div>
    <div>
      <input
        type="text"
        value=""
      />
      <select>
        <option
          selected=""
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </select>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            placeholder="startDate - endDate"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`testRenderExpenseListFiltersWithAltFilters 1`] = `
<body>
  <div>
    <div>
      <input
        type="text"
        value="bills"
      />
      <select>
        <option
          selected=""
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </select>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            placeholder="startDate - endDate"
            type="text"
            value="01/01/1970 - 04/01/1970"
          />
          <button
            aria-label="Close"
            class="react-datepicker__close-icon"
            tabindex="-1"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
